imbi/endpoints/integrations/notifications/processing.py

```diff
@@ -0,0 +1,356 @@
+
+        """
+        notification = await self._get_notification(integration_name,
+                                                    notification_name)
+        self.logger.info('processing notification %s/%s default %r',
+                         integration_name, notification_name,
+                         notification.default_action)
+        if not self._evaluate_filters(notification, body):
+            return
+
+        project = await self._get_project(notification, body)
+        if not project:
+            return
+
+        updates = await self._gather_updates(project, notification, body)
+        if not updates:
+            self.logger.info('no updates found')
+            return
+
+        await self._update_facts(project, updates)
+
+    async def get(self, *, integration_name: str, notification_name: str,
+                  **_kwargs: str) -> None:
+        body: dict[str, str | list[str] | None] = {}
+        for arg, values in self.request.query_arguments.items():
+            # NB ..../get?foo results in {'foo': [b'']} and NOT
+            # {'foo': []} as one might otherwise expect
+            if (len(values) == 0 or len(values) == 1
+                    and values[0] == b''):  # pragma: no branch
+                body[arg] = None
+            elif len(values) == 1:
+                body[arg] = values[0].decode('utf-8')
+            else:
+                raise errors.ApplicationError(
+                    422, 'unprocessable-entity',
+                    'multiple query arguments with same name %r', arg)
+        await self._process_notification(integration_name, notification_name,
+                                         body)
+
+    async def post(self, *, integration_name: str, notification_name: str,
+                   **_kwargs: str) -> None:
+        await self._process_notification(integration_name, notification_name,
+                                         self.get_request_body())
+
+    put = post
+
+    async def _update_facts(self, project, updates) -> None:
+
+        async with self.application.postgres_connector(
+                on_duration=self.on_postgres_timing) as connector:
+            for fact_type_id, value in updates.items():
+                try:
+                    await connector.execute(
+                        'INSERT INTO v1.project_facts'
+                        '            (project_id, fact_type_id, value,'
+                        '             recorded_at, recorded_by)'
+                        '     VALUES (%(project_id)s, %(fact_type_id)s,'
+                        '             %(value)s, CURRENT_TIMESTAMP,'
+                        '             %(username)s)'
+                        '         ON CONFLICT (project_id, fact_type_id)'
+                        '         DO UPDATE'
+                        '        SET value = %(value)s,'
+                        '            recorded_at = CURRENT_TIMESTAMP,'
+                        '            recorded_by = %(username)s',
```

Comment:
may want to wrap the SQL bits in this commit with `re.sub` to remove spaces

Reply:
Won't `re.sub` remove those?

Reply:
It will ... but at the cost of running `re.sub` for _every query_ or placing the query in a constant somewhere else. I really don't like doing the latter since it increases the cognitive complexity. As for the cost of `re.sub()`, luckily the [`re` module includes a compilation cache](https://docs.python.org/3/library/re.html#re.compile) so it won't recreate the RE tables with every call. The cache is limited to something like the most recent 128 entries and we shouldn't hit that limit.

TBH, I question whether the complexity and overhead is worth such a small improvement. We included the `re.sub()` stuff explicitly to remove newlines introduced by multiline string literals from the PostgreSQL query logs. The internal whitespace was just a nice addition. I'm wondering if we should plop the `re.sub` directly in the connector `execute` instead.